### ISSUE TITLE
Would it make sense to use hivemind for distributed training?

### ISSUE THREAD

**Issue by 0xdevalias** [db0/AI-Horde#77]
Splitting this out from the unrelated issue:
>         Not sure if the implementation/etc would be compatible, but here's another distributed StableDiffusion training project a friend recently linked me to:
> 
> - https://github.com/chavinlo/distributed-diffusion
>   - Which seems to build on top of: https://github.com/learning-at-home/hivemind
>     - > Decentralized deep learning in PyTorch. Built to train models on thousands of volunteers across the world.
> 
> _Originally posted by @0xdevalias in https://github.com/db0/AI-Horde/issues/12#issuecomment-1304692808_

Basically, I stumbled across the [hivemind](https://github.com/learning-at-home/hivemind) lib and thought that it could be a useful addition to AI-Horde. I'm not 100% sure how the current distributed process is implemented, but from a quick skim it looked like perhaps you had rolled your own.

Not sure if it's something you already considered and decided against, but wanted to bring it to your attention in case you hadn't seen it before.

**Comment by db0** [comment]
That would take quite a bit of work to onboard to the horde, but it's a promising thing. The problem is that the horde is asynchronous so it could be that the latency would be prohibitive, but I would be willing to consider it, especially if someone sends a PR

**Comment by ndahlquist** [comment]
I think the horde is primarily used for inference, not training. Do any jobs actually do training, or is that planned for the future? If not, it seems like this may provide limited benefit.

**Comment by 0xdevalias** [comment]
Speaking of using hivemind for distributed training/etc, just stumbled across the following on:

- https://rentry.org/sdupdates3

> SD Training Labs is going to conduct the first global public distributed training on November 27th
> 
> - Distributed training information provided to me:
>   - Attempted combination of the compute power of over 40+ peers worldwide to train a finetune of Stable Diffusion with Hivemind
>   - This is an experimental test that is not guaranteed to work
>   - This is a peer-to-peer network.
>     - You can use a VPN to connect
>     - Run inside an isolated container if possible
>     - Developer will try to add code to prevent malicious scripting, but nothing is guaranteed
>   - Current concerns with training like this:
>     - Concern 1 - Poisoning: A node can connect and use a malicious dataset hence affecting the averaged gradients. Similar to a blockchain network, this will only have a small effect on the averaged weights. The larger the amount of malicious nodes connected, the more power they will have on the averaged weights. At the moment we are implementing super basic (and vague) discord account verification.
>     - Concern 2 - RCE: Pickle exploits should not be possible but haven't been tested.
>     - Concern 3 - IP leak & firewall issues: Due to the structure of hivemind, IPs will be seen by other peers. You can avoid this by seting client-only mode, but you will limit the network reach. IPFS should be possible to be used to avoid firewall and NAT issues but doesn't work at the moment

Doing some further googling/etc, it seems that the 'SD Training Labs' discord is:

- https://discord.gg/JHyEskQQcZ

And things are being coordinated in the `#distributed-training` channel, which has a few pinned messages about the training, and links to the following repos:

- https://github.com/learning-at-home/hivemind
  - > Decentralized deep learning in PyTorch. Built to train models on thousands of volunteers across the world.
- https://github.com/chavinlo/distributed-diffusion
  - > Train a Stable Diffusion model over the internet with Hivemind
- https://github.com/bigscience-workshop/petals
  - > Decentralized platform for running 100B+ language models. Work in progress
